designs/nutanix-datacenterconfig-credentialref.md

```diff
@@ -0,0 +1,51 @@
+# Nutanix Datacenter Config Credential Reference
+
+## Introduction
+
+**Problem:** When a management cluster and a workload cluster are present on two different Nutanix Prism Central instances,
+using same credentials to authenticate to both instances is not possible. To remedy this, we need to associate credentials
+with corresponding Nutanix Datacenter Configs.
+
+## Overview of Solution
+With this feature, a user can specify credentials reference (i.e. Name and Namespace of the secret containing the credentials)
+in the Nutanix Datacenter Config. This info will be fetched and used to authenticate to the Nutanix Prism Central instance
+when creating the workload cluster.
+
+### Solution Details
+Extend the `NutanixDatacenterConfig` CRD to include a new field `credentialRef` storing name, namespace, and kind (similar to how
+`credentialsRef` is defined in the `NutanixCluster` CRD).
+
+```go
+// NutanixDatacenterConfigSpec defines the desired state of NutanixDatacenterConfig.
+type NutanixDatacenterConfigSpec struct {
+    ...
+    // CredentialRef is the reference to the secret that contains the credentials
+    // for the Nutanix Prism Central
+    // +optional
+    CredentialRef *NutanixCredentialReference `json:"credentialRef,omitempty"`
+}
+
+// NutanixCredentialKind is the kind of the Nutanix credential
+type NutanixCredentialKind string
+
+// NutanixCredentialReference is the reference to the secret that contains the credentials
+// +kubebuilder:object:generate=true
+type NutanixCredentialReference struct {
+    // Kind of the Nutanix credential
+    // +kubebuilder:validation:Enum=Secret
+    Kind NutanixCredentialKind `json:"kind"`
+    // Name of the credential.
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    Name string `json:"name"`
+    // namespace of the credential.
+    // +optional
+    Namespace string `json:"namespace"`
+}
+```
+
+When creating a workload cluster, the `NutanixDatacenterConfig` referenced in the `Cluster` CRD will be fetched and
+the `credentialRef` field will be used to fetch the secret to get the credentials needed to authenticate with Prism Central
+to create resources needed for the cluster. If the `credentialRef` field is not present, the credentials from the
+cluster (i.e. eksa-system/cluster-name) will be used as default to preserve existing behaviour.
```

Comment:
I like this default behavior. However, what do you think about filling this field also for management clusters? The CLI can do that. And for existing workload clusters, we can add a guard in the controller that fills this with the default secret name if it's not present.

That simplifies the rest of the logic since we don't need to handle two cases everywhere, we just make sure the data exists at the beginning of the reconciliation.

Reply:
That's a good point. I'm game for also filling it in for management clusters. That simplifies the logic elsewhere.

Reply:
@g-gaston @chrisdoherty4  How do we feel about the following? 
```
When creating a workload cluster, the `NutanixDatacenterConfig` referenced in the `Cluster` CRD will be fetched and
the `credentialRef` field will be used to fetch the secret to get the credentials needed to authenticate with Prism Central
to create resources needed for the cluster. If the `credentialRef` field is not present when creating a workload cluster, we can
fail the cluster creation request. During management cluster creation, we can fetch the secrets from the
environment, create the secret manifests, and add the `credentialRef` to the corresponding `NutanixDatacenterConfig` if not present.
This will allow us to assume that `credentialRef` will always be set in the `NutanixDatacenterConfig` in all cases.
```

Reply:
I like it, one minor clarification:
`we can fail the cluster creation request` -> ca we do this in the validation webhook?

@jiayiwang7 this is exactly what we do for snow today, right? Require it for workloads and fill it in from the CLI for management.

Reply:
@thunderboltsid That seems OK to me.

Reply:
Updated to explicitly spell out validation will be done using validation webhook.